src/main/scala/viper/gobra/reporting/StatsCollector.scala

```diff
@@ -0,0 +1,357 @@
+    val memberJson = memberMap.values.map(_.asJson("  ")).mkString(", \n")
+    val json = s"""[
+      |$memberJson
+      |]""".stripMargin
+    json
+  }
+
+  def findGobraMemberByViperMemberName(taskName: String, viperMemberName: String, pkgDir: String): Option[GobraMemberEntry] =
+    viperMemberNameGobraMemberMap.get(viperMemberKey(taskName, viperMemberName, pkgDir))
+      .orElse(viperMemberNameGobraMemberMap.get(viperMemberKey(taskName, viperMemberName)))
+
+  def getWarnings(pkgDir: String, pkg: String, config: Config): Set[String] =
+    memberMap.values
+      .filter(gobraMember => gobraMember.pkgDir == pkgDir && gobraMember.pkg == pkg)
+      .flatMap(gobraMember => {
+        val name = gobraMember.pkg + "." + gobraMember.memberName + gobraMember.args
+        gobraMember.dependencies().flatMap({
+          // Trusted implies abstracted, so we match trusted first
+          case GobraMemberEntry(_, pkg, memberName, _, args, _, true, _) =>
+            Some("Warning: Member " + name + " depends on trusted member " + pkg + "." + memberName + args + "\n")
+          case GobraMemberEntry(_, pkg, memberName, _, args, _, _, true) =>
+            Some("Warning: Member " + name + " depends on abstract member " + pkg + "." + memberName + args + "\n")
+          case GobraMemberEntry(pkgDir, pkg, _, _, _, _, _, _) if !config.inputPackageMap.contains(PackageEntry(pkgDir, pkg)) =>
+            Some("Warning: Depending on imported package, that is not verified: " + pkgDir + " - " + pkg)
+          case _ => None
+        })
+      }).toSet
+
+  def addResult(pkgDir: String,
+                pkg: String,
+                memberName: String,
+                nodeType: String,
+                args: String,
+                viperMember: ViperMemberEntry,
+                isTrusted: Boolean,
+                isAbstract: Boolean): Unit = this.synchronized({
+    val key = gobraMemberKey(pkgDir, pkg, memberName, args)
+
+    val shortViperKey = viperMemberKey(viperMember.taskName, viperMember.memberName, pkgDir)
+    val viperKey = viperMemberKey(viperMember.taskName, viperMember.memberName, pkgDir)
+
+    memberMap.get(key) match {
+      case Some(existing) =>
+        val newViperMembers = existing.viperMembers.get(viperKey).orElse(existing.viperMembers.get(shortViperKey)) match {
+          case Some(existingViperEntry) =>
+            // Merge Viper members that occur multiple times during a task(used for chopper)
+            existing.viperMembers + (viperKey -> ViperMemberEntry(
+              existingViperEntry.memberName,
+              existingViperEntry.taskName,
+              existingViperEntry.time + viperMember.time,
+              existingViperEntry.nodeType,
+              existingViperEntry.dependencies ++ viperMember.dependencies,
+              existingViperEntry.success && viperMember.success,
+              existingViperEntry.cached || viperMember.cached,
+              existingViperEntry.fromImport,
```

Comment:
Here, we are using a synchronized block again. You should try to document why this is required if this is really required. Also, not sure about what kinds of race conditions we can have here, but this is only asserting mutual exclusion when adding a result, although there might still be concurrent accesses e.g. to `memberMap` from other methods.

Reply:
You are correct, I switched to concurrent maps and removed all synchronized blocks